File: Cpp_src/easy/intersection_of_two_linked_lists_160.cc
#include <iostream>
using namespace std;

// Use Two Pointers
// Time Complexity: O(m+n)
// Space Complexity:O(1)
// Runtime: 44 ms, faster than 95.69%
// Memory Usage: 16.7 MB, less than 90.74%

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
    {
        if (headA == NULL || headB == NULL)
        {
            return NULL;
        }

        ListNode *tmp_a = headA;
        ListNode *tmp_b = headB;

        while (tmp_a != tmp_b)
        {
            tmp_a = (tmp_a == NULL) ? headB : tmp_a->next;
            tmp_b = (tmp_b == NULL) ? headA : tmp_b->next;
        }

        return tmp_a;
    }
};

File: Golang_src/easy/intersectionOfTwoLinkedLists_160.go
package easy

// Use two pointers:
// Time Complexity: O(m+n)
// Space Complexity:O(1)
// Runtime: 40 ms, faster than 84.90%
// Memory Usage: 8.6 MB, less than 26.30%
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}

	curA, curB := headA, headB

	for curA != curB {
		if curA == nil {
			curA = headB
		} else {
			curA = curA.Next
		}

		if curB == nil {
			curB = headA
		} else {
			curB = curB.Next
		}
	}

	return curA
}
